Integrate golangci-lint into CI/CD pipeline

ArjunRajpal123 opened the pull request:
## Motivation/Why
Integrate [golangci-lint](https://golangci-lint.run/) into the CI/CD pipeline to ensure code quality and consistency across the project.

## Steps followed
1. Add golangci-lint config with minimal configuration. The config could be based on https://github.com/ccoVeille/golangci-lint-config-examples
2. Extend GitHub Actions Workflow .github/workflows/lint.yaml with golangci-lint job. It should use [golangci-lint-action](https://github.com/marketplace/actions/golangci-lint).
3. Run golangci-lint locally, identify, and fix existing linting issues.
4. Disabled Revive in the golangci-lint config

**Important Notes**
- I initially attempted to follow the [basic golangci-lint config](https://github.com/ccoVeille/golangci-lint-config-examples/tree/main/02-basic), but quickly ran into 34 linting issues (errcheck: 24, staticcheck: 10). To keep this PR focused and manageable, I decided to enable only staticcheck and address all related issues.

- I’ve disabled rule SA1019 (usage of deprecated identifiers) for now, as the only trigger was ast.Object, which is a known case of a deprecated field. Handling this seems better suited for a dedicated follow-up PR.

# Footer
Closes #1250

chavacava commented:
Hi @ArjunRajpal123 thanks for the PR!

IMO we should not disable `revive` but run it with the configuration we used until now (`revive.toml` at the root of the repository) (what do you think @denisvmedia  @alexandear @ccoVeille ?)

ccoVeille commented on .golangci.yml:
```diff
@@ -0,0 +1,39 @@
+
+    - revive
+
+  exclusions:
+    rules:
+      - linters: [staticcheck]
+        # SA1019: Using a deprecated function, variable, constant or field
+        text: "SA1019:"
```
So here is some context

> I’ve disabled rule SA1019 (usage of deprecated identifiers) for now, as the only trigger was ast.Object, which is a known case of a deprecated field. Handling this seems better suited for a dedicated follow-up PR.

There multiple ways to handle this:
- the way you chose
- the fact to disable SA1019 from `linters.settings.staticcheck.check` by using -SA1019
- use //nolint:staticcheck at each place ast.Object is used.

I would have used the last option, and add a TODO if there is something we can do.

ArjunRajpal123 replied:
`//nolint:staticcheck` I like this solution a lot more

denisvmedia replied:
Actually, using `//nolint:staticcheck` is not the same. It will disable the whole linter, not just a particular check.

ccoVeille replied:
Yes, but here it would be disabled for one line only. Instead of disabling SA1019 everywhere.

Staticcheck has some directives.

https://staticcheck.dev/docs/configuration/#line-based-linter-directives

Maybe it works when staticcheck is used through golangci-lint

denisvmedia replied:
Unfortunately, it doesn't wok with golangci-lint. That's a well-known issue. And yes, disabling the whole linter for one line is better than the rule for the whole project. I agree.

ccoVeille commented:
> IMO we should not disable revive but run it with the configuration we used until now

I would say it depends whether we plan to remove revive from the CI.

Here is an option I would suggest if we want to keep revive in the CI:

- define everything we need to have revive with same settings as we had
- but configure golangci-lint-action to disable revive
- this way IDE would use golangci-lint settings.

chavacava commented:
> > IMO we should not disable revive but run it with the configuration we used until now
>  I would say it depends whether we plan to remove revive from the CI.

When I say "we should not disable revive" I mean running it through `golangci-lint`

ccoVeille commented:
> > > IMO we should not disable revive but run it with the configuration we used until now
> >  I would say it depends whether we plan to remove revive from the CI.
> 
> When I say "we should not disable revive" I mean running it through `golangci-lint` 

I wasn't sure, so @ArjunRajpal123 here is what you could do:

enable revive linter via the .golangci.yaml file

Then add a node for linters.settings.revive and adapt the setting you will find in 

https://github.com/mgechev/revive/blob/master/revive.toml

denisvmedia commented:
I'm wondering if it's the same when we use revive directly and revive via golangci-lint. As far as I understand currently we check with the most recent version of revive, while in golangci-lint we won't have the most recent of revive.

ArjunRajpal123 commented:
Some recent Changes to this pr
- [x] used `//nolint:staticcheck` at each place `ast.Object` is used (also verified that each place where apt.Object is used has a todo)
- [x]  enabled revive linter via the .golangci.yaml file
- [x] added a node for linters.settings.revive in [this file](https://github.com/mgechev/revive/blob/master/revive.toml) and adapt the setting you will find in
- [x] removed revive from GHA workflow


Sidenote: found two revive lint errors after enabling revive in `golangci`
<img width="800" alt="image" src="https://github.com/user-attachments/assets/b75fb186-b715-4641-9380-9f80de3c0a2b" />

ccoVeille commented on rule/error_strings_test.go:
```diff
@@ -34,7 +34,8 @@ func TestErrorStringsRule_Configure(t *testing.T) {
 			name:      "Invalid function",
 			arguments: lint.Arguments{"errors."},
-			wantErr:   errors.New("found invalid custom function: errors."),
+			//nolint: revive
```
It should be this

```suggestion
			//nolint:revive
```

Please enable `nolintlint` linter via `.golangci.yaml` 

It would lint, the `//nolint` directives

ccoVeille commented on formatter/formatter_test.go:
```diff
@@ -61,11 +61,15 @@ Warnings:
 		{
 			formatter: &formatter.JSON{},
-			want:      `[{"Severity":"warning","Failure":"test failure","RuleName":"rule","Category":"cat","Position":{"Start":{"Filename":"test.go","Offset":0,"Line":2,"Column":5},"End":{"Filename":"test.go","Offset":0,"Line":2,"Column":10}},"Confidence":0,"ReplacementLine":""}]`,
+			want: `[{"Severity":"warning","Failure":"test failure","RuleName":"rule",` +
+				`"Category":"cat","Position":{"Start":{"Filename":"test.go","Offset":0,"Line":2,"Column":5},` +
+				`"End":{"Filename":"test.go","Offset":0,"Line":2,"Column":10}},"Confidence":0,"ReplacementLine":""}]`,
 		},
 		{
 			formatter: &formatter.NDJSON{},
-			want:      `{"Severity":"warning","Failure":"test failure","RuleName":"rule","Category":"cat","Position":{"Start":{"Filename":"test.go","Offset":0,"Line":2,"Column":5},"End":{"Filename":"test.go","Offset":0,"Line":2,"Column":10}},"Confidence":0,"ReplacementLine":""}`,
+			want: `{"Severity":"warning","Failure":"test failure","RuleName":"rule",` +
+				`"Category":"cat","Position":{"Start":{"Filename":"test.go","Offset":0,"Line":2,"Column":5},` +
+				`"End":{"Filename":"test.go","Offset":0,"Line":2,"Column":10}},"Confidence":0,"ReplacementLine":""}`,
 		},
```
Is this change related to a the line-length rule set to 200?


if it's right, then it means calling revive didn't detect it, while enable the rulevia golangci-lint does  🤔 

Maybe a typo in the revive. toml file?

@ArjunRajpal123 I'm fine with the changes, it improves readability, but I'm unsure why you had to do it.

ArjunRajpal123 replied:
It is related to the line-length rule being set to 200 and I was a bit confused myself as to why it was showing up myself. I am going to look into this more closely.

ccoVeille replied:
It might be a bug in the way revive reads the line length from the config file.

I have locally set up the max line length to 5, and go run . reported nothing 😱

Maybe the config in the current revive.toml is wrong, and is not what the code expects.

Or the opposite, there is small bug in this rule.

Anyway, this should be addressed in another PR.

ccoVeille commented on rule/error_strings_test.go:
```diff
@@ -34,7 +34,8 @@ func TestErrorStringsRule_Configure(t *testing.T) {
 			name:      "Invalid function",
 			arguments: lint.Arguments{"errors."},
-			wantErr:   errors.New("found invalid custom function: errors."),
+			//nolint: revive
+			wantErr: errors.New("found invalid custom function: errors."),
```
Here also, I'm unsure why you had to disable revive rule.

revive was supposed to be enabled on the code

ArjunRajpal123 replied:
<img width="800" alt="image" src="https://github.com/user-attachments/assets/90679c86-6779-4557-a1e0-ce663ad7ca66" />

I ran into the following error above when using golangci+revive, which appears related to the following code linked below. Modifying the `wantErr` wouldn't be the correct fix because that causes tests to fail and modifying the rule below seems out of scope. Also curious as to why this shows up now.

https://github.com/mgechev/revive/blob/f6cc22de4d5c75cf471501affa5522c778078761/rule/error_strings.go#L54-L56

ccoVeille replied:
Here I agree the errors strings shouldn't end with a dot.

But it's a special case, it's the unit tests of errors-strings rule 😬😅

So here we might think about how to avoid adding directives that would allow usto write bad code because we want to test it 😁

ccoVeille replied:
> related to the following code linked below. Modifying the `wantErr` wouldn't be the correct fix because that causes tests to fail and modifying the rule below seems out of scope (disabling felt like the right choice). Also curious as to why this lint-error shows up now.

But you might also round a real bug.

ccoVeille commented on .golangci.yml:
```diff
@@ -0,0 +1,85 @@
+    # Checks Go code for unused constants, variables, functions and types.
+    - unused
+
+    - revive
```
Please enable `nolintlint` linter, as I recommended here

https://github.com/mgechev/revive/pull/1346#discussion_r2090225629

ArjunRajpal123 replied:
I have now enabled this linter

denisvmedia commented on .github/workflows/lint.yaml:
```diff
@@ -48,5 +48,7 @@ jobs:
     - name: Install revive
```
Note, you've removed revive run, but didn't remove the installation step. I put a comment [here](https://github.com/mgechev/revive/pull/1346#issuecomment-2881567252), but it might have got lost. I believe running golangci-lint with the latest bundled revive is not the same as running currently built revive.

ccoVeille replied:
I agree, it's why I asked what was expected here

https://github.com/mgechev/revive/pull/1346#issuecomment-2880000622

denisvmedia commented on .github/workflows/lint.yaml:
```diff
@@ -48,5 +48,7 @@ jobs:
-      run: revive --config revive.toml ./...
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v8
+      with:
+        version: v2.1.6
```
I'm not sure about explicit version here. Will it require manual maintenance in the future?

ArjunRajpal123 replied:
It would require additional manual maintenance, I can set version to `latest` to avoid this
https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#version

ccoVeille replied:
That's a usual way to do.

The idea behind this is to avoid someone opening a small change one thing, won't have to fight with all the change the last version of golangci-lint would report.

The very same way to setup-go action is often set up to use go-1. 24, and not stable, oldstable.

But yes, it requires manual maintenance

denisvmedia replied:
> But yes, it requires manual maintenance

I bet, this will be rarely updated then.

chavacava replied:
IMO a manual update is not a bad thing, by the contrary (we keep the control on the actual version we apply to our code base)

alexandear commented on .golangci.yml:
```diff
@@ -0,0 +1,87 @@
+---
+# golangci-lint configuration file made by @ccoVeille
+# Source: https://github.com/ccoVeille/golangci-lint-config-examples/
+# Author: @ccoVeille
+# License: MIT
+# Variant: 01-defaults
+# Version: v2.0.0
+#
```
@ccoVeille do we need this copyright? During a review process we have changed a lot, so this is not the `01-defaults` configuration from https://github.com/ccoVeille/golangci-lint-config-examples/

ccoVeille replied:
Absolutely not. I added this **to my repository** because:
- I wanted to define the license of my work is MIT
- I have the same headers for all my examples
- some people are copy-pasting my "work" as-is.
- I can track the projects that are using my examples, because I'm curious
- sometimes I contributed to some projects my example files were added, for example I helped a project to migrate to the v2 format.

But then here in the context of this PR:
- this file is the very basic one, so of course it has to be complicated.
- the file diverted and is the results of contributions and feedbacks from many people.
- I chose the MIT license because it permits to be included in any project, no matter its license, and it allows to alter the file in the way, and it can be updated in a way to remove my name.


TL; DR; yes, of course, the whole header can be removed.

The only thing I'm unsure about is if the file doesn't have to keep the MIT license line.

But as revive is licensed under MIT license,
it can be removed.

alexandear commented on .golangci.yml:
```diff
@@ -0,0 +1,87 @@
+#
+version: "2"
+
+linters:
```
A few suggestions from my side as one of the [most active golangci-lint contributors](https://github.com/golangci/golangci-lint/graphs/contributors):

* I'd recommend adding `default: none` to ensure that only explicitly enabled linters are used. See https://github.com/golangci/golangci-lint/blob/deee4e75cd912954538449a2f35d795ea04545ce/.golangci.reference.yml#L15. That way, when a new version of golangci-lint is released, we won't implicitly enable new linters.
* It might be beneficial to enable as many linters as possible from [this list](https://github.com/golangci/golangci-lint/blob/deee4e75cd912954538449a2f35d795ea04545ce/.golangci.reference.yml#L19-L123).
* For easier future maintenance, I suggest sorting the list of enabled linters.
* The `disable` section can be removed, as it's redundant when using `default: none`.

alexandear commented on rule/datarace.go:
```diff
@@ -23,6 +23,7 @@ func (r *DataRaceRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
 		funcResults := funcDecl.Type.Results
 
 		// TODO: ast.Object is deprecated
+		//nolint:staticcheck
```
Maybe better:
```suggestion
		//nolint:staticcheck // TODO: ast.Object is deprecated
```

alexandear replied:
The same below

alexandear commented on rule/error_strings_test.go:
```diff
@@ -34,7 +34,8 @@ func TestErrorStringsRule_Configure(t *testing.T) {
 			name:      "Invalid function",
 			arguments: lint.Arguments{"errors."},
-			wantErr:   errors.New("found invalid custom function: errors."),
+			//nolint:revive
```
I think we should adding an explanation we used `//nolint:revive` here.

ccoVeille replied:
I agree.

We should enforce using an explanation with nolint 

https://github.com/gofr-dev/gofr/blob/dac6bdc735c3b60ccfb17d8e4a799c23e3d7383d/.golangci.yml#L97-L98

alexandear commented on rule/if_return.go:
```diff
@@ -44,7 +44,7 @@ func (w *lintElseError) Visit(node ast.Node) ast.Visitor {
 			}
 			assign, ok := s.Init.(*ast.AssignStmt)
-			if !ok || len(assign.Lhs) != 1 || !(assign.Tok == token.DEFINE || assign.Tok == token.ASSIGN) {
+			if !ok || len(assign.Lhs) != 1 || assign.Tok != token.DEFINE && assign.Tok != token.ASSIGN {
```
I'm not sure if this became more readable. Let's revert.

ccoVeille replied:
I agree

This one is implied by QF0001

Please take a look here, on how to disable the rule that implied this.

https://github.com/gofr-dev/gofr/blob/dac6bdc735c3b60ccfb17d8e4a799c23e3d7383d/.golangci.yml#L141

I had to do it for gofr too

- https://github.com/gofr-dev/gofr/pull/1708

chavacava commented:
A quick note regarding pull requests from first-time contributors: to promote a welcoming atmosphere, we should aim to streamline and simplify the PR workflow. While I'm deeply convinced that having others review your code is a valuable learning experience, we should be cautious not to overwhelm contributors with excessive comments or lengthy discussions on details that can be addressed in subsequent PRs.

alexandear commented:
I've addressed all my comments, fixed some additional issues, and merged the PR.

@ArjunRajpal123 — thank you for your contribution!

Thanks everyone for the review.

If you have suggestions for improving the golangci-lint configuration or the CI/CD setup, please open a separate PR. Don't hesitate to contribute!
